pkg/file/redundancy/getter/getter.go

```diff
@@ -0,0 +1,226 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	size := len(addrs)
+	if fetchTimeout == 0 {
+		fetchTimeout = 30 * time.Second
+	}
+	strategyTimeout := StrategyTimeout
+
+	rsg := &decoder{
+		fetcher:   g,
+		putter:    p,
+		addrs:     addrs,
+		inflight:  make([]atomic.Bool, size),
+		cache:     make(map[string]int, size),
+		waits:     make([]chan struct{}, shardCnt),
+		rsbuf:     make([][]byte, size),
+		ready:     make(chan struct{}, 1),
+		cancel:    cancel,
+		remove:    remove,
+		shardCnt:  shardCnt,
+		parityCnt: size - shardCnt,
+	}
+
+	// after init, cache and wait channels are immutable, need no locking
+	for i := 0; i < shardCnt; i++ {
+		rsg.cache[addrs[i].ByteString()] = i
+		rsg.waits[i] = make(chan struct{})
+	}
+
+	// prefetch chunks according to strategy
+	rsg.wg.Add(1)
+	go func() {
+		rsg.prefetch(ctx, strategy, strict, strategyTimeout, fetchTimeout)
+		rsg.wg.Done()
+	}()
+	return rsg
+}
+
+// Get will call parities and other sibling chunks if the chunk address cannot be retrieved
+// assumes it is called for data shards only
+func (g *decoder) Get(ctx context.Context, addr swarm.Address) (swarm.Chunk, error) {
+	i, ok := g.cache[addr.ByteString()]
+	if !ok {
+		return nil, storage.ErrNotFound
+	}
+	if g.fly(i, true) {
+		g.wg.Add(1)
+		go func() {
+			g.fetch(ctx, i)
+			g.wg.Done()
+		}()
+	}
+	select {
+	case <-g.waits[i]:
+		return swarm.NewChunk(addr, g.rsbuf[i]), nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+// fly commits to retrieve the chunk (fly and land)
+// it marks a chunk as inflight and returns true unless it is already inflight
+// the atomic bool implements a singleflight pattern
+func (g *decoder) fly(i int, up bool) (success bool) {
+	return g.inflight[i].CompareAndSwap(!up, up)
+}
+
+// fetch retrieves a chunk from the underlying storage
+// it must be called asynchonously and only once for each chunk (singleflight pattern)
+// it races with erasure recovery which takes precedence even if it started later
+// due to the fact that erasure recovery could only implement global locking on all shards
+func (g *decoder) fetch(ctx context.Context, i int) {
+	ch, err := g.fetcher.Get(ctx, g.addrs[i])
+	if err != nil {
+		_ = g.fly(i, false) // unset inflight
+		return
+	}
+
+	g.mu.Lock()
+	defer g.mu.Unlock()
+	if i < len(g.waits) {
+		select {
+		case <-g.waits[i]: // if chunk is retrieved, ignore
+			return
+		default:
+		}
+	}
+
+	select {
+	case <-ctx.Done(): // if context is cancelled, ignore
+		_ = g.fly(i, false) // unset inflight
+		return
+	default:
+	}
+
+	//  write chunk to rsbuf and signal waiters
+	g.rsbuf[i] = ch.Data() // save the chunk in the RS buffer
+	if i < len(g.waits) {
+		close(g.waits[i]) // signal that the chunk is retrieved
+	}
+
+	// if all chunks are retrieved, signal ready
+	n := g.fetchedCnt.Add(1)
+	if n == int32(g.shardCnt) {
+		close(g.ready) // signal that just enough chunks are retrieved for decoding
```

Comment:
are we allowed to close ready if the shardCnt includes parity chunks?

Reply:
surely